distributed/semaphore.py

```diff
@@ -0,0 +1,327 @@
+                )
+
+    async def _get_lease(self, client, name, identifier):
+        # We should make sure that the client is already properly registered with the scheduler
+        # otherwise the lease validation will mop up every acquired lease immediately. That's mostly relevant for tests
+        while client not in self.scheduler.clients:
+            await asyncio.sleep(0.0005)  # This value is set somewhat arbitrarily
+
+        result = True
+        if len(self.leases[name]) < self.max_leases[name]:
+            # naive: self.leases[resource] += 1
+            # not naive:
+            self.leases[name].append(identifier)
+            self.leases_per_client[client][name].append(identifier)
+        else:
+            result = False
+        return result
+
+    def _semaphore_exists(self, name):
+        if name not in self.max_leases:
+            logger.exception(f"Semaphore `{name}` not known or already closed.")
+            return False
+        return True
+
+    async def acquire(
+        self, comm=None, name=None, client=None, timeout=None, identifier=None
+    ):
+        with log_errors():
+            if not self._semaphore_exists(name):
+                return
+
+            if isinstance(name, list):
+                name = tuple(name)
+            w = _Watch(timeout)
+            w.start()
+
+            while True:
+                # Reset the event and try to get a release. The event will be set if the state
+                # is changed and helps to identify when it is worth to retry an acquire
+                self.events[name].clear()
+
+                # If we hit the timeout, this cancels the _get_lease
+                future = asyncio.wait_for(
+                    self._get_lease(client, name, identifier), timeout=w.leftover()
+                )
+
+                try:
+                    result = await future
+                except TimeoutError:
+                    result = False
+
+                # If acquiring fails, we wait for the event to be set, i.e. something has
+                # been released and we can try to acquire again (continue loop)
+                if not result:
+                    future = asyncio.wait_for(
+                        self.events[name].wait(), timeout=w.leftover()
+                    )
+                    try:
+                        await future
+                        continue
+                    except TimeoutError:
+                        result = False
+                return result
+
+    def release(self, comm=None, name=None, client=None, identifier=None):
+        with log_errors():
+            if not self._semaphore_exists(name):
+                return
+            if isinstance(name, list):
+                name = tuple(name)
+            if name in self.leases and identifier in self.leases[name]:
+                self._release_value(name, client, identifier)
+            else:
+                raise ValueError(
```

Comment:
do we not care if an exception is raised here or would a warning/log message be more appropriate so as to not block further execution? cc @fjetter

Reply:
This code is triggered by a call from the Client side.  The mechanism that Dask uses to do these sorts of remote calls should serialize the error, send it to the client, and reraise it there.  This would be good to verify though.  Maybe something like the following?

```python
semaphore.acquire()
semaphore.release()

with pytest.raises(ValueError) as info:
    semaphore.release()

assert "already released" in str(info.value)
```

?

Reply:
I think the question is not about how the exception is handled but rahter should we raise an exception or ignore this?

I'm inclined to say this should **not** be an exception. This path is encountered in two scenarios:
1. Releasing too often
2. Semaphore closed already

I could imagine both scenarios to happen in a distributed system due to weird timings and I wouldn't want my application to crash while cleaning up. There are different philosophies for this, I believe and I don't have a very strong opinion here.

Reply:
What is the behavior observed by the user/Client?  Is it the same as they would see with a typical Python Semaphore?  If so, then I'm happy.

Reply:
Described this in the semaphore docstring under notes